rudder-core/src/main/scala/com/normation/rudder/datasources/DataSourceRepository.scala

```diff
@@ -275,5 +281,67 @@ class DataSourceRepoImpl(
 
 }
 
+class PGDataSourceRepository(
+    doobie    : Doobie
+) extends DataSourceRepository with Loggable {
+
+  import doobie._
+
+  override def getAll(): Box[Map[DataSourceId,DataSource]] = {
+    query[DataSource]("""select id, properties from datasources""").vector.map { _.map( ds => (ds.id,ds)).toMap }.attempt.transact(xa).run
+  }
+
+  override def get(sourceId : DataSourceId): Box[Option[DataSource]] = {
+    sql"""select id, properties from datasources where id = ${sourceId.value}""".query[DataSource].option.attempt.transact(xa).run
+  }
+
+  override def save(source : DataSource): Box[DataSource] = {
+    import net.liftweb.json.compactRender
+    val json = compactRender(DataSourceJsonSerializer.serialize(source))
+    val insert = """insert into datasources (id, properties) values (?, ?)"""
+    val update = s"""
+      update datasources
+      set properties= ?
+      where id= ?
+    """
+
+    val sql = for {
+      rowsAffected <- Update[(String,String)](update).run((json, source.id.value))
+      result       <- rowsAffected match {
+                        case 0 =>
+                          logger.warn(s"source ${source.id} is not present in database, create it")
+                          Update[DataSource](insert).run(source)
+                        case 1 => 1.point[ConnectionIO]
+                        case n => throw new RuntimeException(s"Expected 0 or 1 change, not ${n} for ${source.id}")
+                      }
+    } yield {
+      result
+    }
+
+    sql.attempt.transact(xa).run match {
+      case \/-(x)  => Full(source)
+      case -\/(ex) => Failure(s"could not update lastId from database, cause is: ${ex.getMessage}", Full(ex), Empty)
+    }
```

Comment:
There is an auto transform to avoid that boilerplate, you could use:
```
sql.map(_ => source).attempt.transact(xa).run
```
The implicit def is in Doobie object (```xorToBox```)